### ISSUE TITLE
LongIterator.nextValue breaks if not preceded by hasNext

### ISSUE THREAD

**Issue by mtopolnik** [real-logic/Agrona#17]
This is what the `nextValue()` method does [currently](https://github.com/real-
logic/Agrona/blob/master/src/main/java/uk/co/real_logic/agrona/collections/Long2LongHashMap.java#L426):

```java
public long nextValue()
{
    final long entry = entries[index];
    nextIndex();
    return entry;
}
```

It returns whatever is at current index, then moves to the next index. If this method is called in a loop, all map slots (regardless if used or not) will be returned. The loop will also be infinite, wrapping around the slot array.

This probably makes sense for highly optimized usage where API contract is deliberately disobeyed. But perhaps it doesn't suit a public library as much. At least not without severe warnings in the docs :)

**Comment by RichardWarburton** [comment]
IntIterator also suffers from this problem as well actually.

**Comment by mjpt777** [comment]
At some stage we lost the check of the `isValidPosition` field in the the Long2ObjectHashMap that should be checked in the `nextValue()` method. We should fix this and have all map backed structures do similar.

**Comment by RichardWarburton** [comment]
Ok, I've fixed this issue now with e3e9e6f9ab1d32e3a4d6d4e6ebdc09aabe5430e0 c0b52bf6e4aa4f3dae6ff9ce226f5ce92f6a1765 and 0fc526a067a6759164b79219818ee9d63094cf7d. I've also made the Long2LongHashMap an Int2IntHashMap and started auto-generating a Long2Long version.
